presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,231 @@
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("Parallel dynamic filter build can not be duplicated");
+        }
+    }
+
+    private static final int DEFAULT_POSITIONS_LIMIT = 1000;
+
+    private final OperatorContext operatorContext;
+    private final List<Type> types;
+    private final List<Integer> filterChannels;
+    private final TupleDomainSource tupleDomainSource;
+    private final int positionsLimit;
+
+    private Page page;
+    private boolean finishing;
+    private long positionCount;
+    private List<ImmutableList.Builder<Object>> valuesBuilder;
+    private boolean[] nullsAllowed;
+
+    public DynamicFilterSourceOperator(
+            OperatorContext operatorContext,
+            List<Type> types,
+            List<Integer> filterChannels,
+            TupleDomainSource tupleDomainSource)
+    {
+        this(operatorContext, types, filterChannels, tupleDomainSource, DEFAULT_POSITIONS_LIMIT);
+    }
+
+    public DynamicFilterSourceOperator(
+            OperatorContext operatorContext,
+            List<Type> types,
+            List<Integer> filterChannels,
+            TupleDomainSource tupleDomainSource,
+            int positionsLimit)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = ImmutableList.copyOf(types);
+        this.filterChannels = ImmutableList.copyOf(filterChannels);
+        this.valuesBuilder = filterChannels.stream().map(channel -> ImmutableList.builder()).collect(toImmutableList());
+        this.nullsAllowed = new boolean[filterChannels.size()];
+        this.tupleDomainSource = requireNonNull(tupleDomainSource, "tupleDomainSource is null");
+
+        checkArgument(positionsLimit > 0, "positionsLimit must be greater than zero");
+        this.positionsLimit = positionsLimit;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (finishing) {
+            return;
+        }
+        finishing = true;
+        buildTupleDomain();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && page == null;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        this.page = page;
+        processPage();
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        Page page = this.page;
+        this.page = null;
+        return page;
+    }
+
+    private void processPage()
+    {
+        positionCount += page.getPositionCount();
+
+        // we don't want to filter if there're too many values
+        if (positionCount > positionsLimit) {
```

Comment:
Fair point.
We can compile an equivalent of `readNativeType` to avoid expensive if-ology inside that method.
The question is if the average case would be so heavily affected by duplicate values and can't this be overcome by adjusting the limit of rows. I'd avoid expensive lookups to `Set` to see if we've seen value already and based on that, proceed with building `TupleDomain` or abandon it.
Especially that in worse case it'd be O(n*log(n)) complexity vs O(1) (because of linera scan capped on some constant).